File: C++/017_Letter Combinations of a Phone Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.length() == 0)
            return res;
        this->digits = digits;
        dfs(0, "", res);
        return res;
    }
private:
    string digits;
    const string dgal[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    void dfs(int pos, string now, vector<string> &res) {
        if (pos >= digits.length()) {
            res.push_back(now);
            return;
        }
        int j = digits[pos] - '0';
        if (j > 1 && j < 10) {
            for (int i = 0; i < dgal[j].length(); i++)
                dfs(pos + 1, now + dgal[j][i], res);
        }
        else
            dfs(pos + 1, now, res);
    }
};

File: Python/017_Letter Combinations of a Phone Number.py
class Solution(object):
    def __init__(self):
        self.res, self.digits = [], ''
        self.dtol = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']

    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        self.digits, self.res = digits, []
        self._dfs(0, '')
        return self.res

    def _dfs(self, pos, stmp):
        if pos >= len(self.digits):
            if len(stmp) > 0:
                self.res.append(stmp)
            return
        d = ord(self.digits[pos]) - ord('0')
        if 1 < d < 10:
            for i in self.dtol[d]:
                self._dfs(pos + 1, stmp + i)
        else:
            self._dfs(pos + 1, stmp)